ttkwidgets/validated_entries/validators.py

```diff
@@ -0,0 +1,118 @@
+import re
+import sys
+
+from ttkwidgets.utilities import isint, isfloat
+
+
+class Validator:
+    """
+    Base validator class
+
+    Specify the VALIDATE_ON class attribute (defaults to 'all') to change
+    on which tkinter condition the validation will be done.
+
+    Possibilities are :
+        * 'all'
+        * 'none'
+        * 'focus'
+        * 'focusin'
+        * 'focusout'
+        * 'key'
+    """
+    VALIDATE_ON = 'all'
```

Comment:
> 
> 
> This is a widget with a lot of potential I think. I have a few questions, but if I am going out of the intended scope of the widget, please reel me back in.
> 
>     * There is no way that I can see that the programmer can request something of an `is_valid` property or something. Particularly for patterns like e-mails, this seems useful to me, as some input need not be valid when it is entered. For example, entering an e-mail address allows entering any data I like, it seems, requiring the programmer to do another check if it is a form or something like that.

This is a good idea, it's already pretty straightforward to access though, through the `self.entry.VALIDATOR._validate(text)` method. An `is_valid` property is very easy to implement, and I will implement it tonight, after work

>     * Maybe a visual cue could be implemented to indicate the validity of the input? Doing so on the entry itself is complex in `ttk`, especially if custom themes are used, but maybe a checkmark or something similar?

This widget is intended to be as barebones as possible. It's a validating entry, nothing more. Visual cues can be added by subclassing the widgets, but there is so many options for visual cues that it doesn't make sense to implement it in the widget itself.

>     * Separating the `Validator` and the `ValidatorEntry` feels a bit Java-style, but I understand the reasoning. Perhaps add an option for simply passing a regex to a `ValidatorEntry` to allow quick creation if a `Validator` need only be used once?

It is a bit Java-style but the intention is to have a clear separation between the data and the view, Integrating validation on the widget itself imo would blur the lines and make this whole thing less modular

>     * From my basic testing, it seems that `Validators` may not be combined or removed. Is this correct?

I intend to make it possible to have multiple validators for one widget (hence this particular architecture) and I will work on that tonight as well.

Reply:
It does belong to the validator though, because the validator is responsible for generating the appropriate dict for the entry, not to mention that this is already a kwarg when instanciating the entry. I can absolutely check that kwargs are not superseeded by the validator though.

Reply:
You can pass in the 'validate' kwarg to the entry to overwrite the validate_on attribute of the validator.